### ISSUE TITLE
Train GAN from scratch is so difficult!

### ISSUE THREAD

**Issue by feature-space-move** [csxmli2016/MARCONet#30]
I'm trying to train GAN part w/o loading pretrain and the size is same with MARCONet. Encoder and SR part load pretrain in this repo and freeze. 
It seems like GAN don't output binary mask? Maybe training is not enough, or my training crash?
Show you something：
1. train loss
<img width="2078" height="1264" alt="Image" src="https://github.com/user-attachments/assets/1541429c-e70f-4a48-a123-34b456f5519c" />
2. inference
GAN use weights I trained and other model use public in this repo. 
<img width="2100" height="1008" alt="Image" src="https://github.com/user-attachments/assets/e6d65fcb-2f80-4b05-9ce8-a51eeb605103" />

Hope for reply. Thanks!

**Comment by csxmli2016** [comment]
> I'm trying to train GAN part w/o loading pretrain and the size is same with MARCONet. Encoder and SR part load pretrain in this repo and freeze. It seems like GAN don't output binary mask? Maybe training is not enough, or my training crash? Show you something：
> 
> 1. train loss
> 
> <img alt="Image" width="2000" height="1264" src="https://private-user-images.githubusercontent.com/108510408/521658352-1541429c-e70f-4a48-a123-34b456f5519c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQ4MTAzMDEsIm5iZiI6MTc2NDgxMDAwMSwicGF0aCI6Ii8xMDg1MTA0MDgvNTIxNjU4MzUyLTE1NDE0MjljLWU3MGYtNGE0OC1hMTIzLTM0YjQ1NmY1NTE5Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMjA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTIwNFQwMTAwMDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNGZhNzVjYjc5MWY1YTU0NDdlNjhhYzQ3YmEzOWI0MjQwOTY1NzM1YWM2ZDBkMWUzNzM0ZmU5MzBhM2JkYmQwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.knMShnqjyV_AFROy4I_uXofXH_2D33GpTznMzz8Tptw"> 2. inference GAN use weights I trained and other model use public in this repo. <img alt="Image" width="2000" height="1008" src="https://private-user-images.githubusercontent.com/108510408/521659500-e6d65fcb-2f80-4b05-9ce8-a51eeb605103.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQ4MTAzMDEsIm5iZiI6MTc2NDgxMDAwMSwicGF0aCI6Ii8xMDg1MTA0MDgvNTIxNjU5NTAwLWU2ZDY1ZmNiLTJmODAtNGIwNS05Y2U4LWE1MWVlYjYwNTEwMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMjA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTIwNFQwMTAwMDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01YjdkNWM4MTM0MzI3MmQxYzZhOTlmN2EyZjYzOWQ3MGU0MDNjY2RiN2Q4ODI4YjA4OGZjMjMwZmM4ZmJjZGYyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7zhxmU30otOI3WP5BknLMqodmGBS3vTGJlRHFMcsyIE">
> Hope for reply. Thanks!

You mean train the generative prior from scratch? If so, you can remove the Encoder and SR modules and use recognition loss and adversarial loss. In this training process, you can observe the generated character structure through TensorBoard. All black means crash. 

<img width="861" height="343" alt="Image" src="https://github.com/user-attachments/assets/33b9c127-a767-4b07-a500-6f31ea23da26" />

**Comment by feature-space-move** [comment]
> > I'm trying to train GAN part w/o loading pretrain and the size is same with MARCONet. Encoder and SR part load pretrain in this repo and freeze. It seems like GAN don't output binary mask? Maybe training is not enough, or my training crash? Show you something：
> > 
> > 1. train loss
> > 
> > <img alt="Image" width="2000" height="1264" src="https://private-user-images.githubusercontent.com/108510408/521658352-1541429c-e70f-4a48-a123-34b456f5519c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQ4MTAzMDEsIm5iZiI6MTc2NDgxMDAwMSwicGF0aCI6Ii8xMDg1MTA0MDgvNTIxNjU4MzUyLTE1NDE0MjljLWU3MGYtNGE0OC1hMTIzLTM0YjQ1NmY1NTE5Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMjA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTIwNFQwMTAwMDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNGZhNzVjYjc5MWY1YTU0NDdlNjhhYzQ3YmEzOWI0MjQwOTY1NzM1YWM2ZDBkMWUzNzM0ZmU5MzBhM2JkYmQwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.knMShnqjyV_AFROy4I_uXofXH_2D33GpTznMzz8Tptw"> 2. inference GAN use weights I trained and other model use public in this repo. <img alt="Image" width="2000" height="1008" src="https://private-user-images.githubusercontent.com/108510408/521659500-e6d65fcb-2f80-4b05-9ce8-a51eeb605103.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQ4MTAzMDEsIm5iZiI6MTc2NDgxMDAwMSwicGF0aCI6Ii8xMDg1MTA0MDgvNTIxNjU5NTAwLWU2ZDY1ZmNiLTJmODAtNGIwNS05Y2U4LWE1MWVlYjYwNTEwMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMjA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTIwNFQwMTAwMDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01YjdkNWM4MTM0MzI3MmQxYzZhOTlmN2EyZjYzOWQ3MGU0MDNjY2RiN2Q4ODI4YjA4OGZjMjMwZmM4ZmJjZGYyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7zhxmU30otOI3WP5BknLMqodmGBS3vTGJlRHFMcsyIE">
> > Hope for reply. Thanks!
> 
> You mean train the generative prior from scratch? If so, you can remove the Encoder and SR modules and use recognition loss and adversarial loss. In this training process, you can observe the generated character structure through TensorBoard. All black means crash.
> 
> <img alt="Image" width="861" height="343" src="https://private-user-images.githubusercontent.com/26357039/522182057-33b9c127-a767-4b07-a500-6f31ea23da26.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQ4MzU3MzksIm5iZiI6MTc2NDgzNTQzOSwicGF0aCI6Ii8yNjM1NzAzOS81MjIxODIwNTctMzNiOWMxMjctYTc2Ny00YjA3LWE1MDAtNmYzMWVhMjNkYTI2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjA0VDA4MDM1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI5NWRlNDg2MWJjMmRiZDJjMjM4NWM3NzE4MjEyMjRkNGE0MzBmZThkNDVlODYwZjIwOTYxMDY5YzQwMzkwMmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.acTxlkLr9tbHbw7sr-HuY8dYqHcIy7uTNEtzqV4ZgG4">

Thanks for replying so timely and I can hardly express my gratitude!
- Yes,  I am training the generative prior from scratch and freeze other model to ensure reproducing results, and then I could reduce the num of params in GAN part(maybe all MARCONet, I need a mini and powerful one).
- What about I train Encoder/GAN/SR model sequentially? Is this the best solution to get a mini one?
- Your remark "All black means crash" gave me a very important reference. My training I mentioned above is still going, and the GAN part seems to be waking up. Should I be happy about this? Please watch what I attached below.

<img width="2806" height="544" alt="Image" src="https://github.com/user-attachments/assets/26249087-b785-4aee-941b-a7a6eb3ef366" />

<img width="2090" height="1024" alt="Image" src="https://github.com/user-attachments/assets/5f409e02-98d8-471e-b06c-fc895af0b6ba" />

**Comment by csxmli2016** [comment]
@Suncheng2022 about 2, you should train the GAN first, and then the Encoder and SR. for 3, this is a sign that GAN is getting better, just waiting for convergence.

**Comment by feature-space-move** [comment]
> [@Suncheng2022](https://github.com/Suncheng2022) about 2, you should train the GAN first, and then the Encoder and SR. for 3, this is a sign that GAN is getting better, just waiting for convergence.

How about jointly training the `encoder`, `GAN`, and `SR` components with reduced parameters(maybe long time)? Currently, their parameter counts are 17M, 7M, and 4M, respectively. 
I noticed that training the GAN alone doesn't seem to work well due to the lack of a proper reference for the style vector `w`, because the `w` must match `encoder`.

**Comment by csxmli2016** [comment]
> > [@Suncheng2022](https://github.com/Suncheng2022) about 2, you should train the GAN first, and then the Encoder and SR. for 3, this is a sign that GAN is getting better, just waiting for convergence.
> 
> How about jointly training the `encoder`, `GAN`, and `SR` components with reduced parameters(maybe long time)? Currently, their parameter counts are 17M, 7M, and 4M, respectively. I noticed that training the GAN alone doesn't seem to work well due to the lack of a proper reference for the style vector `w`, because the `w` must match `encoder`.

Ideally, it could work when training them for a very long time. But actually, this is more like using a pre-trained generative GAN for face or other image restoration. Even with a long training time, its performance is not equal to that one using pre-trained GAN.

**Comment by feature-space-move** [comment]
> > > [@Suncheng2022](https://github.com/Suncheng2022) about 2, you should train the GAN first, and then the Encoder and SR. for 3, this is a sign that GAN is getting better, just waiting for convergence.
> > 
> > 
> > How about jointly training the `encoder`, `GAN`, and `SR` components with reduced parameters(maybe long time)? Currently, their parameter counts are 17M, 7M, and 4M, respectively. I noticed that training the GAN alone doesn't seem to work well due to the lack of a proper reference for the style vector `w`, because the `w` must match `encoder`.
> 
> Ideally, it could work when training them for a very long time. But actually, this is more like using a pre-trained generative GAN for face or other image restoration. Even with a long training time, its performance is not equal to that one using pre-trained GAN.

I believe that you are right, but I need a mini yet equally powerful GAN. Besides training from scratch, is there any way to get a samller one?

**Comment by csxmli2016** [comment]
@feature-space-move Thanks for your interest. More details are confidential. We welcome the opportunity for a deeper commercial discussion
